src/main/scala/io/opentargets/etl/backend/Interactions.scala

```diff
@@ -0,0 +1,390 @@
+    def generateMapping(rnaCentral: DataFrame, humanMapping: DataFrame): DataFrame = {
+      val targets = df
+        .withColumn("proteins", coalesce(col("proteinAnnotations.accessions"), array()))
+        .select("id", "proteins")
+      val humanMappingDF = humanMapping.transformHumanMapping
+      val rnaMappingDF = rnaCentral.transformRnacentral
+
+      val mappingHuman = (
+        targets
+          .join(humanMappingDF, Seq("id"), "left")
+          .withColumn(
+            "mapped_id_list",
+            when(col("mapping_list").isNull, col("proteins"))
+              .otherwise(array_union(col("proteins"), col("mapping_list")))
+          )
+          .select("id", "mapped_id_list")
+          .distinct
+          .withColumnRenamed(
+            "id",
+            "gene_id"
+          )
+        )
+
+      val mappingExplode =
+        mappingHuman.withColumn("mapped_id", explode(col("mapped_id_list"))).drop("mapped_id_list")
+
+      val mapping = mappingExplode.union(rnaMappingDF)
+
+      mapping
+
+    }
+
+    /** generate the interactions from Intact resource
+      * df is the implicit dataframe: Intact dataframe
+      * @param target Dataframe with list of ensembl_id, protein_id
+      * @param rnacentral Dataframe with the rna_id, ensembl_id
+      * @param humanmapping dataframe with human_id, ensembl_id
+      * @return a DataFrame
+      */
+    def generateIntacts(
+        targets: DataFrame,
+        rnacentral: DataFrame,
+        humanmapping: DataFrame
+    ): DataFrame = {
+      val mappingDF = targets.generateMapping(rnacentral, humanmapping)
+      val intactInteractions = df.generateInteractions(mappingDF)
+      intactInteractions
+    }
+
+    /** generate the interactions from Strings resource.
+      * df is the implicit dataframe: Strings dataframe
+      * @param ensproteins Dataframe with the protein_id, ensembl_id
+      * @return a DataFrame
+      */
+    def generateStrings(ensproteins: DataFrame): DataFrame = {
+      val mapping = ensproteins.withColumnRenamed("protein_id", "mapped_id").distinct
+      val stringInteractions = df
+        .generateInteractions(mapping)
+        .filter(col("evidences.evidence_score") > 0)
+      stringInteractions
+    }
+
+    // Common procedure to transform the Dataframes in common Interaction entries.
+
+    /** generate a string truncated at - or _ char.
+      * Eg. "URS123-2_992   return URS123"
+      * @param s string with possible _ or _ char
+      * @return a string
+      */
+    val getCodeFcn = udf { s: String => s.split("_")(0).split("-")(0) }
+
+    /** generate the interactions from a common Dataframe schema
+      * If causalInteraction is true -> swap (A, B) and add to the dataframe
+      * @param mappingInfo Dataframe with mapping_id, ensembl_id
+      * @return a DataFrame
+      */
+    def generateInteractions(mappingInfo: DataFrame): DataFrame = {
+
+      val interactions = df
+        .withColumn("intA", col("interactorA.id"))
```

Comment:
If these `withColumn` calls are only renaming columns we should use `.select(col(interactorA.id.as("intA"))` as the withColumn call has a risk of a StackOverflowError: https://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html#withColumn(colName:String,col:org.apache.spark.sql.Column):org.apache.spark.sql.DataFrame. 
Another option is to bundle it later in the `selectExpr`

Reply:
Done